src/ott/solvers/iterative_linear.py

```diff
@@ -0,0 +1,303 @@
+    return children, aux_data
+
+  def init_state(
+      self,
+      init_linear_pb: linear_problem.LinearProblem,
+      init_prob_state: NamedTuple,
+      rng: jax.random.PRNGKeyArray,
+  ) -> IterLinState:
+    """Initialize the state of the IterativeLinearSolver.
+
+    Args:
+    init_linear_pb: initialization for linear OT problem
+    init_prob_state: initalization of prob_state
+    rng: Random key for low-rank initializers. Only used when
+    :attr:`warm_start` is `False`.
+
+    Returns:
+    The initial IterLinState .
+    """
+    linear_state = self.linear_ot_solver(init_linear_pb)
+    num_iter = self.max_iterations
+    if self.store_inner_errors:
+      errors = -jnp.ones((num_iter, self.linear_ot_solver.outer_iterations))
+    else:
+      errors = None
+
+    return IterLinState(
+        costs=-jnp.ones((num_iter,)),
+        linear_convergence=-jnp.ones((num_iter,)),
+        linear_pb=init_linear_pb,
+        linear_state=linear_state,
+        prob_state=init_prob_state,
+        rngs=jax.random.split(rng, num_iter),
+        errors=errors,
+    )
+
+  def __call__(
+      self,
+      next_linear_pb: Callable[
+          [IterLinState, jax.random.PRNGKeyArray],
+          linear_problem.LinearProblem,
+      ],
+      new_pstate_cost: Callable[
+          [
+              IterLinState,
+              linear_problem.LinearProblem,
+              "LinearOutput",
+              jax.random.PRNGKeyArray,
+          ],
+          Tuple[NamedTuple, float],
+      ],
+      init_linear_pb: Optional[linear_problem.LinearProblem] = None,
+      init_prob_state: Optional[NamedTuple] = None,
+      init_state: Optional[IterLinState] = None,
+      rng: Optional[jax.random.PRNGKeyArray] = None,
+  ) -> IterLinState:
+    """Run the generic solver.
+
+    Args:
+    next_linear_pb: a user-defined function that takes the current state a
+      random key and outputs a new linear problem
+    new_pstate_cost: a user-defined function that takes the current state,
+      the new linear problem, new linear solution, a random key and and
+      outputs a new state
+    init_linear_pb: the initial linear problem to be solved
+    init_prob_state: initialization for the problem state
+    init_state: allows the user to directly define the initial state.
+      If provided init_prob and init-aux state are ignored
+    rng: Random number key.
+
+    Returns:
+    Last solver state.
+    """
+    if rng is None:
+      rng = utils.default_prng_key(rng)
+
+    if (
+        init_linear_pb is None or init_prob_state is None
+    ) and init_state is None:
+      raise ValueError(
+          "If initial solver state is None, init_linear_pb and \
+            init_prob_state should be provided"
+      )
+    if init_state is None:
+      init_state = self.init_state(init_linear_pb, init_prob_state, rng)
+
+    return iterations(self, init_state, next_linear_pb, new_pstate_cost)
+
+  @property
+  def warm_start(self) -> bool:
```

Comment:
I don't think this is needed, can be removed (in the original paper, I haven't seen warm starts).